executor/sample.go

```diff
@@ -0,0 +1,365 @@
+			}
+			cols = append(cols, tableCol)
+		}
+	}
+	// Schema columns contain _tidb_rowid, append extra handle column info.
+	if len(cols) < len(schemaCols) && schemaCols[len(schemaCols)-1].ID == model.ExtraHandleID {
+		extraHandle := model.NewExtraHandleColInfo()
+		extraHandle.Offset = len(cols)
+		tableCol := &table.Column{ColumnInfo: extraHandle}
+		colMap[model.ExtraHandleID] = decoder.Column{
+			Col: tableCol,
+		}
+		cols = append(cols, tableCol)
+	}
+	return cols, colMap, nil
+}
+
+func (s *tableRegionSampler) scanFirstKVForEachRange(ranges []kv.KeyRange,
+	fn func(handle kv.Handle, value []byte) error) error {
+	ver := kv.Version{Ver: s.startTS}
+	snap := s.ctx.GetStore().GetSnapshot(ver)
+	concurrency := sampleMethodRegionConcurrency
+	if len(ranges) < concurrency {
+		concurrency = len(ranges)
+	}
+
+	fetchers := make([]*sampleFetcher, concurrency)
+	for i := 0; i < len(ranges); i++ {
+		fetchers[i] = &sampleFetcher{
+			workerID:    i,
+			concurrency: concurrency,
+			kvChan:      make(chan sampleKV),
+			snapshot:    snap,
+			ranges:      ranges,
+		}
+		go fetchers[i].run()
+	}
+	syncer := sampleSyncer{
+		fetchers:   fetchers,
+		totalCount: len(ranges),
+		consumeFn:  fn,
+	}
+	return syncer.sync()
+}
+
+func (s *tableRegionSampler) resetRowMap() {
+	if s.rowMap == nil {
+		colLen := len(s.schema.Columns)
+		s.rowMap = make(map[int64]types.Datum, colLen)
+		return
+	}
+	for id := range s.rowMap {
+		delete(s.rowMap, id)
+	}
+}
+
+func (s *tableRegionSampler) finished() bool {
+	return s.isFinished
+}
+
+type sampleKV struct {
+	handle kv.Handle
+	value  []byte
+}
+
+type sampleFetcher struct {
+	workerID    int
+	concurrency int
+	kvChan      chan sampleKV
+	err         error
+	snapshot    kv.Snapshot
+	ranges      []kv.KeyRange
+}
+
+func (s *sampleFetcher) run() {
+	defer close(s.kvChan)
+	for i, r := range s.ranges {
+		if i%s.concurrency != s.workerID {
+			continue
+		}
+		it, err := s.snapshot.Iter(r.StartKey, r.EndKey)
+		if err != nil {
+			s.err = err
+			return
+		}
+		for it.Valid() {
+			if !tablecodec.IsRecordKey(it.Key()) {
+				if err = it.Next(); err != nil {
+					s.err = err
+					return
+				}
+				continue
+			}
+			handle, err := tablecodec.DecodeRowKey(it.Key())
+			if err != nil {
+				s.err = err
+				return
+			}
+			s.kvChan <- sampleKV{handle: handle, value: it.Value()}
+			break
+		}
+	}
+}
+
+type sampleSyncer struct {
+	fetchers   []*sampleFetcher
+	totalCount int
+	consumeFn  func(handle kv.Handle, value []byte) error
+}
+
+func (s *sampleSyncer) sync() error {
+	defer func() {
+		for _, f := range s.fetchers {
+			// Cleanup channels to terminate fetcher goroutines.
+			for _, ok := <-f.kvChan; ok; {
+			}
+		}
+	}()
+	for i := 0; i < s.totalCount; i++ {
+		f := s.fetchers[i%len(s.fetchers)]
+		select {
+		case v, ok := <-f.kvChan:
+			if ok {
+				err := s.consumeFn(v.handle, v.value)
+				if err != nil {
+					return err
+				}
+			}
+			if f.err != nil {
+				return f.err
+			}
```

Comment:
if ok is false, that means the channel is closed by defer function? I think we should break the select channel here.